### ISSUE TITLE
az disk list doesn't list a disk that definitely exists

### ISSUE THREAD

**Issue by ondrejbudai** [Azure/azure-cli#17123]
## Describe the bug

`az disk list` doesn't list a disk that definitely exists. I can confirm that by using `az disk show`.

## To Reproduce:
```bash
az disk show --resource-group sharing-research --name test_disk1_8f9bb5540a564309ab5749c8fce60c96
az disk list --resource-group sharing-research
```

## Logs:
```bash
$ az disk show --resource-group sharing-research --name test_disk1_8f9bb5540a564309ab5749c8fce60c96
{
  "burstingEnabled": null,
  "creationData": {
    "createOption": "Copy",
    "galleryImageReference": null,
    "imageReference": null,
    "logicalSectorSize": null,
    "sourceResourceId": "/subscriptions/8d026bb1-2a65-454d-a88f-c896db94c4f8/resourceGroups/sharing-research/providers/Microsoft.Compute/snapshots/test0_0_dqsundlpiwo",
    "sourceUniqueId": "920a5f2e-55c5-4cbf-9fd1-55bf869542e4",
    "sourceUri": null,
    "storageAccountId": null,
    "uploadSizeBytes": null
  },
  "diskAccessId": null,
  "diskIopsReadOnly": null,
  "diskIopsReadWrite": 500,
  "diskMBpsReadOnly": null,
  "diskMBpsReadWrite": 60,
  "diskSizeBytes": 2147483648,
  "diskSizeGb": 2,
  "diskState": "Unattached",
  "encryption": {
    "diskEncryptionSetId": null,
    "type": "EncryptionAtRestWithPlatformKey"
  },
  "encryptionSettingsCollection": null,
  "extendedLocation": null,
  "hyperVGeneration": "V1",
  "id": "/subscriptions/8d026bb1-2a65-454d-a88f-c896db94c4f8/resourceGroups/sharing-research/providers/Microsoft.Compute/disks/test_disk1_8f9bb5540a564309ab5749c8fce60c96",
  "location": "westeurope",
  "managedBy": null,
  "managedByExtended": null,
  "maxShares": null,
  "name": "test_disk1_8f9bb5540a564309ab5749c8fce60c96",
  "networkAccessPolicy": "AllowAll",
  "osType": "Linux",
  "provisioningState": "Succeeded",
  "purchasePlan": null,
  "resourceGroup": "sharing-research",
  "shareInfo": null,
  "sku": {
    "name": "Standard_LRS",
    "tier": "Standard"
  },
  "tags": {},
  "tier": null,
  "timeCreated": "2021-02-26T06:17:56.867885+00:00",
  "type": "Microsoft.Compute/disks",
  "uniqueId": "8f9bb554-0a56-4309-ab57-49c8fce60c96",
  "zones": null
}
$ az disk list --resource-group sharing-research
[]
```

Logs with `--debug`: https://gist.github.com/ondrejbudai/ba27559a40a07290b17202f9680f0709

## Environment Summary
This is inside Ubuntu container on Fedora:
```
Linux-5.10.16-200.fc33.x86_64-x86_64-with-debian-bullseye-sid
Python 3.6.10
Installer: DEB

azure-cli 2.19.1
```

This is on Fedora:
```
Linux-5.10.16-200.fc33.x86_64-x86_64-with-glibc2.32
Python 3.9.1
Installer: RPM

azure-cli 2.19.1
```

Both environments have the same issue.

**Comment by yungezz** [comment]
hi @qwordy could you pls have a look? thanks

**Comment by qwordy** [comment]
If
```
azure.core.pipeline.policies._universal: Request URL: 'https://management.azure.com/subscriptions/8d026bb1-2a65-454d-a88f-c896db94c4f8/resourceGroups/sharing-research/providers/Microsoft.Compute/disks/test_disk1_8f9bb5540a564309ab5749c8fce60c96?api-version=2020-09-30'
azure.core.pipeline.policies._universal: Request method: 'GET'
```
returns the disk, then
```
azure.core.pipeline.policies._universal: Request URL: 'https://management.azure.com/subscriptions/8d026bb1-2a65-454d-a88f-c896db94c4f8/resourceGroups/sharing-research/providers/Microsoft.Compute/disks?api-version=2020-09-30'
azure.core.pipeline.policies._universal: Request method: 'GET'
```
should return it.
@ondrejbudai Could you confirm the disk is not deleted after the first call?

**Comment by ondrejbudai** [comment]
No, the disk is still there. Interestingly, after 3 days, this works as expected, I can see the disk using `az disk list`.

**Comment by ondrejbudai** [comment]
I tried it again: I created a new disk and this time, I can see it in `az disk list` but I cannot see it in `az resource list`.

Regarding the disk from Friday: I can see it in both `az disk list` and `az resource list` now (both didn't show it on Friday).

**Comment by qwordy** [comment]
@ondrejbudai It acts like a lucky draw. If it happens again, comment here.

**Comment by ondrejbudai** [comment]
I've just remembered that I dealt with the same issue like a half year ago when I implemented a test that created an azure VM from an image and then deleted it including all related resources. I used the "resource list" API call directly that time and I had issues that the disk couldn't be found many times (randomly).

I, therefore, suspect that this might be an actual API issue - may I ask you where I can report Azure API issues?

**Comment by qwordy** [comment]
Added a service attention label.
Hi service team, it is reported that https://docs.microsoft.com/en-us/rest/api/resources/resources/list or https://docs.microsoft.com/en-us/rest/api/resources/resources/listbyresourcegroup can't list the expected disk resource. Could you help take a look? @ondrejbudai can provide more details of how to reproduce it.

**Comment by amjads1** [comment]
@Vaibhav-Agar Can you please look into this issue?

**Comment by Drewm3** [comment]
I believe this is Azure Resource Manager (ARM) behavior. ARM is a multi-master, multi-instance, multi-region service. So it is always possible to see something for a few moments even after a delete or not see something for a few moments right after create. 

Let me redirect to the ARM team to see if they have guidance on how to minimize this behavior, but I do not know if it is possible to eliminate the behavior completely with this type of distributed system.

**Comment by qwordy** [comment]
According to @Drewm3, The data will be eventually consistent in a distributed system. But we don't know what the exact latency is.

**Comment by ondrejbudai** [comment]
Thanks for looking into this! May I ask if there is docs on how to create a VM and then delete it with all associated resources including disks?

**Comment by qwordy** [comment]
@ondrejbudai `vm delete` only deletes VM. We need to run other commands to delete other resources. Another way is delete the whole resource group. I think your suggestion is promising. I am also tired of deleting resources one by one.

**Comment by cgrubesol** [comment]
I am seeing this same issue. I can see my disks in the console and I can see them using `az disk show` but I cannot see the disks using `az disk list` or `az resource list`

This is VERY problematic. Disks are expensive and it doesn't seem like we have a way via the CLI to know if we have any disks in order to implement a full cleanup. As a result we have left dangling disks and been charged for them. 

Right now I am not seeing the API as being "eventually consistent". Of course I have not tested over days, but definitely hours. If it takes days to be eventually consistent I don't know if that will really help us.

With this issue I don't really see how Azure deployments can be involved in any DevOps pipelines because I cannot automate  my cleanup processes.

Has anyone identified a workaround or does Microsoft have a recommendation how to properly cleanup a VM including its disks?

**Comment by havlicekp** [comment]
Same here. I've created a VM ~25 minutes ago. `az disk list` doesn't list it:
`PS C:\> az disk list[]`

**Comment by bondib** [comment]
I'm encountering the same issue when trying to find a disk (right after creation and sometimes for hours after creation) with "az disk list --resource-group RG-NAME".
Maybe it will help someone and not sure if this workaround always works but when I remove the "--resource-group" filter - I do find the disk (together with all other disks in the subscription :-(   )

**Comment by robinmholt** [comment]
Is there any progress on this.  Today, I created a virtual machine using 'az cli'.  I immediately tried to do an 'az resource list...' and do not see the -Disk-OS, but I do see an additional disk created with the same deployment.  The disk does show up using 'az disk list'

For my purposes, I have found that I need to do both an 'az resource list' and an 'az disk list' to get the full set of resources for my VM and then feed that to the 'az delete'.

For other's reference, here is the script I am using to delete all resources.  In my case, we have a naming convention that the hostname for the vm is used as a prefix on all of its resources.  It occasionally takes 2 tries to actually get them all, so I put it in a while loop checking up to 5 times.

`
TRIES=0

while [ ${TRIES} -lt 5 ]; do
  RESOURCES=$( (az resource list --subscription "${SUBSCRIPTION}" --resource-group "${RG}" | jq -r -c ".[] | select (.name | contains(\"${HOSTNAME}\")) | .id"; az disk list --subscription "${SUBSCRIPTION}" --resource-group "${RG}" | jq -r -c ".[] | select (.name | contains(\"${HOSTNAME}\")) | .id") | sort -uf | xargs echo )
  if [ ! -z "${RESOURCES}" ]; then
    echo "Deleting $( echo ${RESOURCES} | wc -w) resources."
    az resource delete --ids ${RESOURCES} --verbose
  else
    break
  fi
  TRIES=$(( TRIES + 1 ))
done
`

**Comment by mario-shippo** [comment]
Do we have any progress on this issue? I'm still seeing the same problem after an year and all required updates...?!?
One interesting fact is that when I'm running the az disk list from M1 Mac I can see all disks, but when doing from inside a Linux Gitlab Runner it is listing like 1048 lines of data returned and then nothing else... like 40% of disks.
This is quite concerning issue.
